### PR TITLE
Remove assertions from storage code

### PR DESCRIPTION
**Pull request by Jinming-Hu**
closes https://github.com/Azure/azure-sdk-for-cpp/issues/2183

# Pull Request Checklist

Please leverage this checklist as a reminder to address commonly occurring feedback when submitting a pull request to make sure your PR can be reviewed quickly:

See the detailed list in the [contributing guide](https://github.com/Azure/azure-sdk-for-cpp/blob/master/CONTRIBUTING.md#pull-requests).

- [x] [C++ Guidelines](https://azure.github.io/azure-sdk/cpp_introduction.html)
- [x] Doxygen docs
- [x] Unit tests
- [x] No unwanted commits/changes
- [x] Descriptive title/description
  - [x] PR is single purpose
  - [x] Related issue listed
- [x] Comments in source
- [x] No typos
- [x] Update changelog
- [x] Not work-in-progress
- [x] External references or docs updated
- [x] Self review of PR done
- [x] Any breaking changes?

### CODE REVIEW

File: sdk/storage/azure-storage-files-datalake/src/datalake_file_client.cpp [hunk 1]
```diff
@@ -157,13 +157,11 @@ namespace Azure { namespace Storage { namespace Files { namespace DataLake {
...
-    if (options.ContentHash.HasValue())
+    if (options.ContentHash.HasValue()
+        && options.ContentHash.Value().Algorithm == HashAlgorithm::Md5)
     {
-      AZURE_ASSERT_MSG(
-          options.ContentHash.Value().Algorithm == HashAlgorithm::Md5,
-          "This operation only supports MD5 content hash.");
+      protocolLayerOptions.ContentMd5 = options.ContentHash;
     }
-    protocolLayerOptions.ContentMd5 = options.ContentHash;
```
Comment (RickWinter) [inline sdk/storage/azure-storage-files-datalake/src/datalake_file_client.cpp:166]: Why is this not going to the server and letting the server reject it?

cc: @JeffreyRichter, @tg-msft

File: sdk/storage/azure-storage-files-shares/src/share_file_client.cpp [hunk 2]
```diff
@@ -164,11 +164,9 @@ namespace Azure { namespace Storage { namespace Files { namespace Shares {
     {
       protocolLayerOptions.FileContentDisposition = options.HttpHeaders.ContentDisposition;
     }
-    if (!options.HttpHeaders.ContentHash.Value.empty())
+    if (!options.HttpHeaders.ContentHash.Value.empty()
+        && options.HttpHeaders.ContentHash.Algorithm == HashAlgorithm::Md5)
     {
-      AZURE_ASSERT_MSG(
-          options.HttpHeaders.ContentHash.Algorithm == HashAlgorithm::Md5,
-          "This operation only supports MD5 content hash");
```
Comment (RickWinter) [inline sdk/storage/azure-storage-files-shares/src/share_file_client.cpp:171]: Ditto on letting the server reject the request if the options are invalid.

**Review by RickWinter — CHANGES_REQUESTED** [review]

File: sdk/storage/azure-storage-files-datalake/src/datalake_file_client.cpp (hunk 1 above)
Comment (JeffreyRichter) [reply sdk/storage/azure-storage-files-datalake/src/datalake_file_client.cpp:166]: Yes, I think the header should just be set regardless and the service deals with it.

Comment (Jinming-Hu) [reply sdk/storage/azure-storage-files-datalake/src/datalake_file_client.cpp:166]: > Why is this not going to the server and letting the server reject it?

1. Server won't reject it, server ignores this header.

2. Protocol layer code is auto-generated based on swagger (or something works like swagger). There's no way to add some headers that are not defined in swagger.

File: sdk/storage/azure-storage-files-shares/src/share_file_client.cpp (hunk 2 above)
Comment (Jinming-Hu) [reply sdk/storage/azure-storage-files-shares/src/share_file_client.cpp:171]: > Ditto on letting the server reject the request if the options are invalid.

Server won't reject it, server ignores this header.

Protocol layer code is auto-generated based on swagger (or something works like swagger). There's no way to add some headers that are not defined in swagger.

File: sdk/storage/azure-storage-files-datalake/src/datalake_file_client.cpp (hunk 1 above)
Comment (Jinming-Hu) [reply sdk/storage/azure-storage-files-datalake/src/datalake_file_client.cpp:166]: > Yes, I think the header should just be set regardless and the service deals with it.

And what the HTTP header name would be?
What if service team decides to use a different HTTP header name when they support this feature?

**Pull request closed without merging**